### PR TITLE
Refactor document endpoint take 4

### PR DESCRIPTION
**Pull request by wonrax**
## Type

- [ ] Feature
- [x] Bug fix
- [ ] CI/CD
- [ ] Documentation
- [ ] Other

## Description

This PR groups all document query related resolvers to one, and
conditionally fetch metadata or proof status depending on whether the
graphql client requests for them.

Removed the document field value validation in the prover since
they're already validated on the graphql layer.

Refactored prover operations to be more efficient and sound.

### CODE REVIEW

File: packages/serverless/src/apollo/app/document.ts [hunk 1]
```diff
@@ -85,147 +75,146 @@ export const typeDefsDocument = gql`
...
-  type ListDocumentWithMetadataResponse {
-    document: DocumentResponse!
-    metadata: MetadataDocumentResponse!
+  type DocumentResponse {
+    docId: String!
+    document: JSON
+    createdAt: Date
+    updatedAt: Date
+    metadata: MetadataDocumentResponse
     proofStatus: String
```
Comment (SangTran-127) [inline packages/serverless/src/apollo/app/document.ts:84]: Using Enum for proofStatus

File: packages/serverless/src/apollo/app/document.ts [hunk 2]
```diff
@@ -10,48 +13,42 @@ import {
...
   TDocumentModificationResponse,
   TDocumentHistoryResponse,
+  TPaginationReturn,
+  TDocumentFindResponse,
+  PERMISSION_DEFAULT_VALUE,
 } from '@zkdb/common';
 
-import { DEFAULT_PAGINATION } from '@common';
 import { Permission } from '@zkdb/permission';
+import { GraphqlHelper } from 'src/helper/graphql';
```
Comment (SangTran-127) [inline packages/serverless/src/apollo/app/document.ts:27]: danger import when build js dont know where 'src/' to resolve

File: packages/serverless/src/apollo/app/document.ts [hunk 3]
```diff
@@ -85,147 +75,146 @@ export const typeDefsDocument = gql`
...
 
-// Query
-const findDocument = authorizeWrapper<
+const documentFind = authorizeWrapper<
   TDocumentFindRequest,
-  TDocumentRecordNullable | null
->(JOI_DOCUMENT_FIND_REQUEST, async (_root: unknown, args, ctx) => {
-  const document = await withTransaction((session) =>
-    Document.findDocument(
+  TPaginationReturn<TDocumentFindResponse[]>
```
Comment (SangTran-127) [inline packages/serverless/src/apollo/app/document.ts:144]:
```suggestion
  TDocumentFindResponse = TPaginationReturn<Data>
```

File: packages/common/src/validation/index.ts [hunk 4]
```diff
@@ -132,8 +138,8 @@ export const documentField = Joi.object<TDocumentField, true>()
       case 'Int64':
       case 'UInt64': {
         if (typeof raw.value !== 'string' || typeof raw.value !== 'number') {
-          return helpers.error(
-            'Value must be a number or a string representing a number'
+          throw Error(
+            `Field ${name} of kind ${kind}'s value must be a number or a string representing a number`
           );
```
Comment (SangTran-127) [inline packages/common/src/validation/index.ts:143]: should be throw new Error()

File: packages/common/src/validation/index.ts [hunk 5]
```diff
@@ -171,19 +181,19 @@ export const documentRecord = Joi.object<
   true
 >().custom((raw, helpers) => {
   if (typeof raw !== 'object' || raw === null) {
-    return helpers.error('Document must be an object');
+    throw Error('Document must be an object');
   }
```
Comment (SangTran-127) [inline packages/common/src/validation/index.ts:186]: throw new Error

File: packages/common/src/validation/index.ts [hunk 6]
```diff
@@ -152,13 +160,15 @@ export const documentField = Joi.object<TDocumentField, true>()
...
+          throw new Error(
+            `Field ${name} of kind ${kind} validation error: ${error}`
+          );
         }
         field = { kind, value };
         break;
       }
       default:
-        return helpers.error(`Unsupported kind: ${kind}`);
+        throw Error(`Unsupported kind: ${kind}`);
```
[the remark this answers is not in the document]
Comment (wonrax) [reply packages/common/src/validation/index.ts:171]: fixed

File: packages/common/src/validation/index.ts [hunk 7]
```diff
@@ -143,7 +149,9 @@ export const documentField = Joi.object<TDocumentField, true>()
           break;
         } catch (e) {
           if (e instanceof Error) {
-            return helpers.error(e.message);
+            throw Error(
+              `Field ${name} of kind ${kind} validation error: ${e.message}`
```
[the remark this answers is not in the document]
Comment (wonrax) [reply packages/common/src/validation/index.ts:153]: fixed

File: packages/common/src/validation/index.ts (hunk 4 above)
Comment (wonrax) [reply packages/common/src/validation/index.ts:143]: fixed

File: packages/serverless/src/apollo/app/document.ts (hunk 1 above)
Comment (wonrax) [reply packages/serverless/src/apollo/app/document.ts]: fixed

File: packages/serverless/src/apollo/app/document.ts (hunk 2 above)
Comment (wonrax) [reply packages/serverless/src/apollo/app/document.ts:28]: fixed

File: packages/serverless/src/apollo/app/document.ts (hunk 3 above)
Comment (wonrax) [reply packages/serverless/src/apollo/app/document.ts]: fixed

File: packages/serverless/src/apollo/app/document.ts [hunk 8]
```diff
@@ -85,147 +75,146 @@ export const typeDefsDocument = gql`
...
+    'data',
+    'metadata',
+  ]);
+  const includesProofStatus = GraphqlHelper.checkRequestedFieldExist(info, [
+    'data',
+    'proofStatus',
+  ]);
+
+  return await withTransaction(async (session) => {
+    let [listDocument, numTotalDocument] = await Document.query(
```
[the remark this answers is not in the document]
Comment (wonrax) [reply packages/serverless/src/apollo/app/document.ts]: fixed

**Pull request merged**